src/tactic/chain_trans.lean

```diff
@@ -0,0 +1,164 @@
+specify whether the assumption is a strict ordering
+(i.e. `<`) or non-strict ordering (i.e. `≤`). -/
+@[derive [inhabited, decidable_eq]]
+inductive edge
+| lt | le
+
+instance : has_to_string edge :=
+⟨ λ e, match e with
+      | edge.lt := "lt"
+      | edge.le := "le"
+      end ⟩
+
+meta instance edge.has_to_format : has_to_format edge :=
+⟨ λ e, to_fmt $ to_string e ⟩
+
+open native
+
+/--
+Place `lt` edges first in edge list.
+-/
+meta def sort_edges (ls : list (expr × edge × expr)) : list (expr × edge × expr) :=
+let ⟨xs₀, xs₁⟩ := ls.partition (λ e : _ × _ × _, e.2.1 = edge.lt) in
+xs₀ ++ xs₁
+
+/-- A graph, represented as adjacency lists, where edges are
+expressions of type `α` and where an edge between `u` and `v` contains
+a proof that `u ≤ v` or `u < v`.
+
+Each edge is stored as a triple (proof, label, target) in
+the list of edges of its source. The label is `lt` or `le` and
+the proof establishes the appropriate relation between the source
+and the target of the edge.
+
+The list of edges of each vertex is sorted so as to keep the
+edges labelled `lt` first in the list. This ensures that,
+when traversing the graph, paths that prove a stronger relation
+are going to be considered first. -/
+meta def graph := native.rb_lmap expr (expr × edge × expr)
+
+/--
+Construct a `graph` from a list of edges. Edges are represented as
+tuples `(proof, label, source, target)`.
+-/
+meta def graph.mk (es : list (expr × edge × expr × expr)) : graph :=
+rb_map.map sort_edges $ es.foldl (λ (g : graph) ⟨pr,e,x,y⟩, g.insert x (pr,e,y)) (rb_lmap.mk _ _)
+
+meta instance graph.has_to_format : has_to_tactic_format graph :=
+by delta graph; apply_instance
+
+private meta def dfs_trans' (g : graph) (r : ref expr_set) (v : expr) :
+  expr → list (edge × expr) → tactic (option expr)
+| x hs := do
+  vs ← read_ref r,
+  if vs.contains x then pure none
+  else if v = x then do
+    (h :: hs) ← pure hs | some <$> mk_mapp ``le_refl [none, none, x],
+    (some ∘ prod.snd) <$> hs.mfoldl (λ ⟨e',h'⟩ ⟨e, h⟩,
+              match e, e' with
+              | edge.lt, edge.lt := prod.mk edge.lt <$> mk_app ``lt_trans [h, h']
+              | edge.lt, edge.le := prod.mk edge.lt <$> mk_app ``lt_of_lt_of_le [h, h']
+              | edge.le, edge.lt := prod.mk edge.lt <$> mk_app ``lt_of_le_of_lt [h, h']
+              | edge.le, edge.le := prod.mk edge.le <$> mk_app ``le_trans [h, h']
+              end) h
+  else do
+    vs ← write_ref r (vs.insert x),
+    option_t.run $ (g.find x).mfirst $ λ ⟨h',e',y⟩, option_t.mk $ dfs_trans' y ((e', h') :: hs)
+
+/--
+Depth first search in a graph of ordered relation. Finds a proof
+of `v ≤ v'` or `v < v'`, whichever is strongest and true.
+-/
+meta def dfs_trans (g : graph) (v v' : expr) : tactic expr :=
+using_new_ref mk_expr_set $ λ r, dfs_trans' g r v' v [] >>= option.to_monad
+
+end chain_trans
+
+namespace interactive
+open chain_trans
+
+/--
+Prove a goal of the shape `x ≤ y` or `x < y` by finding assumptions of
```

Comment:
It would be nice if the tactic called (a variant of) `cc` if the goal is of the form `x = y`, to make the user interface more uniform.

Reply:
I'm considering using monotonicity instead of congruence and already this makes for a significantly more complex tactic. I think making it future work might be best.